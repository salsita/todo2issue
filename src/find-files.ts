import globby from 'globby'

export function findFiles(root: string, patterns: string[]): Promise<string[]> {
  return globby(patterns, {
    cwd: root,
    expandDirectories: false,
    gitignore: true
  })
}
