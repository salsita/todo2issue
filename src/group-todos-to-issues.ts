import { Issue, Todo } from './model'

export function groupTodosToIssues(todos: Todo[]): Issue[] {
  const issues: Issue[] = []
  const issuesByNumber = new Map<number, Issue>()
  todos
    .filter((todo) => !!todo.issueNumber)
    .forEach((todo) => {
      let issue = issuesByNumber.get(todo.issueNumber!)
      if (!issue) {
        issue = { todos: [], issueNumber: todo.issueNumber }
        issuesByNumber.set(todo.issueNumber!, issue)
        issues.push(issue)
      }

      issue.todos.push(todo)
    })

  const issuesByText = new Map<string, Issue>()

  Array.from(issuesByNumber.values()).forEach((issue) => {
    issue.todos.forEach((todo) => issuesByText.set(todo.text, issue))
  })

  todos
    .filter((todo) => !todo.issueNumber)
    .forEach((todo) => {
      let issue = issuesByText.get(todo.text)
      if (!issue) {
        issue = { todos: [] }
        issuesByText.set(todo.text, issue)
        issues.push(issue)
      }
      issue.todos.push(todo)
    })

  return issues
}
