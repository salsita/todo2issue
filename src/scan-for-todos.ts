import { readFile } from 'fs/promises'
import { extname, resolve } from 'path'
import { parse } from 'leasot'
import { Todo } from './model'

const ISSUE_NUMBER_PATTERN = /^#([0-9]+)$/

export async function scanForTodos(root: string, files: string[]): Promise<Todo[]> {
  const todos: Todo[] = []
  for (const filename of files) {
    const contents = await readFile(resolve(root, filename), 'utf8')
    const todoComments = parse(contents, { filename, extension: extname(filename) })

    todos.push(
      ...todoComments.map((todoComment) => {
        const { line, text, tag, ref } = todoComment
        const [, issueNumber] = ISSUE_NUMBER_PATTERN.exec(ref) || []
        const todo = {
          filename,
          line,
          tag,
          text,
          issueNumber: issueNumber ? Number(issueNumber) : undefined,
        }
        return todo
      }),
    )
  }
  return todos
}
