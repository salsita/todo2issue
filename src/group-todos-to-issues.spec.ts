import { groupTodosToIssues } from './group-todos-to-issues'
import { scanForTodos } from './scan-for-todos'
import { resolve } from 'path'
import { findFiles } from './find-files'

describe('groupTodosToIssues',() => {
  it('groups', async () => {
    const files = await findFiles(resolve(__dirname, '../sample'), ['**/*.[tj]s?(x)'])
    const todos = await scanForTodos(resolve(__dirname, '../sample'), files)
    const issues = groupTodosToIssues(todos)
    expect(issues).toMatchSnapshot()
  })
})
