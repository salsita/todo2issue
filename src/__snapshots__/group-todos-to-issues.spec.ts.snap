// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groupTodosToIssues groups 1`] = `
Array [
  Object {
    "issueNumber": 123,
    "todos": Array [
      Object {
        "filename": "src/index.ts",
        "issueNumber": 123,
        "line": 2,
        "tag": "TODO",
        "text": "Single line todo comment with reference",
      },
      Object {
        "filename": "src/index.ts",
        "issueNumber": 123,
        "line": 4,
        "tag": "TODO",
        "text": "Same reference different comment",
      },
      Object {
        "filename": "src/index.ts",
        "issueNumber": undefined,
        "line": 5,
        "tag": "TODO",
        "text": "Same reference different comment",
      },
    ],
  },
  Object {
    "issueNumber": 124,
    "todos": Array [
      Object {
        "filename": "src/index.ts",
        "issueNumber": 124,
        "line": 12,
        "tag": "TODO",
        "text": "Multiline todo",
      },
      Object {
        "filename": "src/index.ts",
        "issueNumber": undefined,
        "line": 7,
        "tag": "TODO",
        "text": "Multiline todo",
      },
      Object {
        "filename": "src/helpers/helper.ts",
        "issueNumber": undefined,
        "line": 4,
        "tag": "TODO",
        "text": "Multiline todo",
      },
    ],
  },
  Object {
    "issueNumber": 125,
    "todos": Array [
      Object {
        "filename": "src/index.ts",
        "issueNumber": 125,
        "line": 19,
        "tag": "TODO",
        "text": "Multiple single",
      },
      Object {
        "filename": "src/helpers/helper.ts",
        "issueNumber": 125,
        "line": 7,
        "tag": "TODO",
        "text": "Related issue",
      },
      Object {
        "filename": "src/index.ts",
        "issueNumber": undefined,
        "line": 15,
        "tag": "TODO",
        "text": "Multiple single",
      },
    ],
  },
  Object {
    "todos": Array [
      Object {
        "filename": "src/index.ts",
        "issueNumber": undefined,
        "line": 1,
        "tag": "TODO",
        "text": "Single line todo comment",
      },
    ],
  },
  Object {
    "todos": Array [
      Object {
        "filename": "src/index.ts",
        "issueNumber": undefined,
        "line": 16,
        "tag": "TODO",
        "text": "line comments",
      },
    ],
  },
  Object {
    "todos": Array [
      Object {
        "filename": "src/index.ts",
        "issueNumber": undefined,
        "line": 20,
        "tag": "TODO",
        "text": "line comments with reference",
      },
    ],
  },
  Object {
    "todos": Array [
      Object {
        "filename": "src/index.ts",
        "issueNumber": undefined,
        "line": 22,
        "tag": "TODO",
        "text": "With colon",
      },
      Object {
        "filename": "src/index.ts",
        "issueNumber": undefined,
        "line": 24,
        "tag": "TODO",
        "text": "With colon",
      },
      Object {
        "filename": "src/index.ts",
        "issueNumber": undefined,
        "line": 26,
        "tag": "TODO",
        "text": "With colon",
      },
    ],
  },
  Object {
    "todos": Array [
      Object {
        "filename": "src/helpers/helper.ts",
        "issueNumber": undefined,
        "line": 1,
        "tag": "TODO",
        "text": "This one is from a sub-folder",
      },
    ],
  },
]
`;
